tests/view_tests/tests/test_defaults.py

```diff
@@ -129,3 +128,15 @@ def test_custom_templates_wrong(self):
 
         with self.assertRaises(TemplateDoesNotExist):
             server_error(request, template_name='nonexistent')
+
+    def test_all_responses_are_a_valid_html_document(self):
+        """
+        Responses from error pages are a valid HTML document (#30521).
+        """
+        request = self.request_factory.get('/')
+        for response in [page_not_found(request, None), server_error(request), bad_request(request, None), permission_denied(request, None)]:
+            self.assertTrue(response.content.startswith(b'<!DOCTYPE html>'), msg=response)
+            self.assertContains(response, b'<html lang="en">', status_code=response.status_code)
+            self.assertContains(response, b'<body>', status_code=response.status_code)
+            self.assertContains(response, b'</body>', status_code=response.status_code)
+            self.assertContains(response, b'</html>', status_code=response.status_code)
```

Comment:
I would use `assertIn` instead of `self.assertContains` and `self.subTest`, e.g.
```python
def test_error_pages(self):
    request = self.request_factory.get('/')
    for response, title in (
        (bad_request(request, Exception()), b'Bad Request (400)'),
        (permission_denied(request, Exception()), b'403 Forbidden'),
        (page_not_found(request, Http404()), b'Not Found'),
        (server_error(request), b'Server Error (500)'),
    ):
        with self.subTest(title=title):
            self.assertIn(b'<!DOCTYPE html>', response.content)
            self.assertIn(b'<html lang="en">', response.content)
            self.assertIn(b'<head>', response.content)
            self.assertIn(b'<title>%s</title>' % title, response.content)
            self.assertIn(b'<body>', response.content)
```